### PR TITLE
[string] introduce StringWriter class

### PR DESCRIPTION
**Pull request by bukepo**
This commit moves methods of String out of the header file to save code
size.

Since most use cases of String writers ignores errors, this commit
changes the return of writers to the Writer object itself for chained
writes.

### CODE REVIEW

**Review by abtink — APPROVED** [review]
This looks great. Thanks @bukepo. 👍

File: src/core/common/string.hpp [hunk 1]
```diff
@@ -90,76 +90,64 @@ const char *StringFind(const char *aString, char aChar);
...
+private:
+    char mBuffer[kSize];
 };
 
-/**
- * This class defines a fixed-size string.
- *
- */
-template <uint16_t SIZE> class String : private StringBase
+class StringWriter
```
Comment (jwhui) [inline src/core/common/string.hpp:118]: nit: add doxygen?

File: src/core/common/string.cpp [hunk 2]
```diff
@@ -72,30 +73,56 @@ bool StringEndsWith(const char *aString, char aChar)
...
 {
-    Error error = kErrorNone;
-    int   len;
+    mBuffer[0] = 0;
+}
 
-    len = vsnprintf(aBuffer + aLength, aSize - aLength, aFormat, aArgs);
+StringWriter &StringWriter::Clear(void)
+{
+    mBuffer[0] = 0;
```
Comment (jwhui) [inline src/core/common/string.cpp:86]: nit: to be more explicit
```suggestion
    mBuffer[0] = '\0';
```

File: src/core/common/string.hpp [hunk 3]
```diff
@@ -90,76 +90,64 @@ const char *StringFind(const char *aString, char aChar);
...
+     * @retval  true    Buffer is sufficient.
+     * @retval  false   Buffer is insufficient.
      *
      */
-    void Clear(void)
-    {
-        mBuffer[0] = 0;
-        mLength    = 0;
-    }
+    bool IsSufficient(void) const { return mLength < mSize; }
```
Comment (jwhui) [inline src/core/common/string.hpp:150]: nit: a user of this method may find something like `IsTruncated()` more intuitive?

**Review by jwhui — APPROVED** [review]

**Pull request merged**